### PR TITLE
UI: Fixes search bar and debounce for search

### PR DESCRIPTION
**Pull request by pratap0007**
This patch fixes following
    
   - Previously search bar was showing up searched text after
      a delay(~300 ms),now search bar gets reflected immediate
      while searching a resource
    
  - Previously debounce was not working as expected and now it is fixed
      with help of `use-debounce` npm package and removes useDebounce component
      from utils directory, now users will  get filtered resources just
      after 500 ms when they stop typing in the search bar
    
Closes: #252 

Signed-off-by: Shiv Verma <shverma@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if functionality changed/added)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/main/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: ui/src/containers/Search/index.tsx [hunk 1]
```diff
@@ -18,11 +25,22 @@ const Search: React.FC = () => {
...
-  const inputRef = React.useRef<HTMLInputElement>(null);
+  const onSearchChange = () => {
+    const currentInput = inputRef.current;
+    assert(currentInput);
+    setValue(currentInput.value);
+    debounced(currentInput.value);
+  };
 
-  const onSearchChange = useDebounce(resources.search, 400);
+  const inputRef = React.useRef<HTMLInputElement>(null);
```
Comment (PuneetPunamiya) [inline ui/src/containers/Search/index.tsx:43]: Can you move this above the `onSearchChange` function

File: ui/src/containers/Search/index.tsx [hunk 2]
```diff
@@ -18,11 +25,22 @@ const Search: React.FC = () => {
     }, 20);
   });
 
-  const { resources } = useMst();
+  const debounced = useDebouncedCallback(
```
Comment (PuneetPunamiya) [inline ui/src/containers/Search/index.tsx:28]: Could you add a comment here what this function does ??

File: ui/src/containers/Search/index.tsx [hunk 3]
```diff
@@ -18,11 +25,22 @@ const Search: React.FC = () => {
...
+    (value: string) => {
+      resources.setSearch(value);
+    },
+    // delay in ms
```
Comment (PuneetPunamiya) [inline ui/src/containers/Search/index.tsx:32]: nit: 
```suggestion
    // Delay in ms
```

File: ui/src/containers/Search/index.tsx [hunk 4]
```diff
@@ -18,11 +25,22 @@ const Search: React.FC = () => {
...
+    500
+  );
 
-  const inputRef = React.useRef<HTMLInputElement>(null);
+  const onSearchChange = () => {
+    const currentInput = inputRef.current;
+    assert(currentInput);
```
Comment (PuneetPunamiya) [inline ui/src/containers/Search/index.tsx:38]: Can we just assert like this 
```suggestion
    assert(inputRed.current);
```

**Review by PuneetPunamiya — COMMENTED** [review]
Thanks for the patch!
Could you please address the minor comments

File: ui/src/containers/Search/index.tsx (hunk 2 above)
Comment (pratap0007) [reply ui/src/containers/Search/index.tsx:28]: yes sure

File: ui/src/containers/App/__snapshots__/App.test.tsx.snap [hunk 5]
```diff
@@ -29,13 +29,13 @@ exports[`App should render the component correctly and match the snapshot 1`] =
                           <div className=\\"pf-l-grid\\">
                             <GridItem span={11}>
                               <div className=\\"pf-l-grid__item pf-m-11-col\\">
-                                <Search>
-                                  <ForwardRef value=\\"\\" type=\\"search\\" onChange={[Function: onChange]} onKeyPress={[Function: onSearchKeyPress]} aria-label=\\"text input example\\" placeholder=\\"Search for resources...\\" spellCheck=\\"false\\" className=\\"hub-search\\">
-                                    <TextInputBase value=\\"\\" type=\\"search\\" onChange={[Function: onChange]} onKeyPress={[Function: onSearchKeyPress]} aria-label=\\"text input example\\" placeholder=\\"Search for resources...\\" spellCheck=\\"false\\" className=\\"hub-search\\" innerRef={{...}} isRequired={false} validated=\\"default\\" isDisabled={false} isReadOnly={false} isLeftTruncated={false}>
+                                <Memo(wrappedComponent)>
```
Comment (PuneetPunamiya) [inline ui/src/containers/App/__snapshots__/App.test.tsx.snap:32]: Is this change in test expected here ??

File: ui/src/containers/Search/index.tsx [hunk 6]
```diff
@@ -1,28 +1,48 @@
...
+  /* Debounce callback
+     This function delay the invoking func for 500ms after
+     when users stop searching
+  */
+  const debounced = useDebouncedCallback(
+    (value: string) => {
+      resources.setSearch(value);
+    },
+    // Delay in ms
+    500
```
Comment (PuneetPunamiya) [inline ui/src/containers/Search/index.tsx:37]: Why the delay in search is 500 ??
Where the user data which he has typed for searching will be stored till 500ms are done? Should that be stored in the localstorage ??

File: ui/src/containers/App/__snapshots__/App.test.tsx.snap (hunk 5 above)
Comment (pratap0007) [reply ui/src/containers/App/__snapshots__/App.test.tsx.snap:32]: yes

File: ui/src/containers/Search/index.tsx (hunk 6 above)
Comment (pratap0007) [reply ui/src/containers/Search/index.tsx:37]: it's a custom delay , I kept it 500 ms (I thought 500 ms is fine ,if you think any other time would be fine we can change it).
search data is fetching from the `search box` itself

Comment (PuneetPunamiya) [reply ui/src/containers/Search/index.tsx:37]: Does it depend on the number of resources in the store ??

Comment (pratap0007) [reply ui/src/containers/Search/index.tsx:37]: nope , it's a delay to used for debounce (any function  will be execute only after that delay), In our case store get update (filter function to be called) only just after 500 ms when user stop searching

**Review by vinamra28 — COMMENTED** [review]
/lgtm

**Pull request merged**